### ISSUE TITLE
Unable to create alternative XHRBackend

### ISSUE THREAD

**Issue by thelgevold** [angular/angular#11178]
**I'm submitting a ...**  (check one with "x")

```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
In the angular.io project we rely on a fake xhr implementation to simulate a web api in our samples.  The implementation can be found here: https://github.com/angular/in-memory-web-api/blob/master/src/in-memory-backend.service.ts#L120

To short circuit the injection of `XHRBackend`, we provide the fake backend like this:
`{ provide: XHRBackend, useClass: InMemoryBackendService }`

This approach works well for JIT compiled samples, but for some reason this does not work when doing AOT compilation. 

The compile error I am seeing is:

```
Error: Error at /Users/tor/Helgevold-Consulting-Dev/angular.io/public/docs/_examples/toh-6/ts/app/app.module.ngfactory.ts:219:67: Supplied parameters do not match any signature of call target.
Error at /Users/tor/Helgevold-Consulting-Dev/angular.io/public/docs/_examples/toh-6/ts/app/app.module.ngfactory.ts:227:75: Argument of type 'InMemoryBackendService' is not assignable to parameter of type 'XHRBackend'.
  Property '_browserXHR' is missing in type 'InMemoryBackendService'.
```

It complains that `_browserXHR` is not defined in the class. 

I tried defining `_browserXHR`, but that leads me to the new error:

```
Argument of type 'InMemoryBackendService' is not assignable to parameter of type 'XHRBackend'.
  Types have separate declarations of a private property '_browserXHR'.
```

I am not sure why this doesn't AOT compile. Why is it comparing internal (private) properties between the two implementations?

Thanks!

**Comment by thelgevold** [comment]
@chuckjaz To me this error message seems a bit contradictory. Initially it says the field is missing, but once I define it, it's says it's a 'duplicate'. I might be missing something though...

**Comment by vicb** [comment]
Would marking `InMemoryBackendService` as `@Injectable` help ? (the other option w/o modifying the source code is to use a factory)

**Comment by robwormald** [comment]
This isn't a codegen problem, per se, its a Typescript thing that's exposed by codegen (AoT doesn't do any typechecking itself)

See https://github.com/Microsoft/TypeScript/issues/471 for the mechanics of why this is the way it is.

The reason this surfaces for XHRBackend and not in other places is that there's no HttpBackend type (its an interface) and the default setup for Http is `constructor(backend:XHRBackend)` - we should probably create an abstract HttpBackend type, use that for the constructor param and `{ provide:HttpBackend, useClass: XHRBackend }` in the Http module.

**Comment by wardbell** [comment]
@robwormald  Your response confuses me. For one,

**Comment by robwormald** [comment]
I didn't say XHRBackend was an interface, I said *ConnectionBackend* was an interface. That's the base type (its actually abstract, not an interface) that DI is depending on. However, because we use a factory to provide Http - https://github.com/angular/angular/blob/master/modules/%40angular/http/src/http_module.ts#L31-L33 consumers are forced to override XHRBackend (which is incorrect)

Thus, the generated code calls the factory:
```ts
get _Http_31():any {
    if ((this.__Http_31 == (null as any))) { (this.__Http_31 = import6.httpFactory(this._XHRBackend_29,this._RequestOptions_30)); }
    return this.__Http_31;
  }
```
And then you run into the issue with TS i linked to, because there's no way to implement it.

We *should* be doing something like in the HttpModule
```
[ Http, { provide: ConnectionBackend, useValue: XHRBackend } ]
```
Those who want to override the backend should then
```
{ provide: ConnectionBackend, useClass: MyFakeConnectionBackend }
```

tldr: XHRBackend is the implementation of ConnectionBackend. Consumers of the Http module should *not* override XHRBackend, they should override ConnectionBackend. That's currently not possible as wired up.

In the interim, ` { provide: Http, useFactory: _myCustomHttpFactory, deps: [ MyBackend, DefaultOptions } ` should get around the problem.

**Comment by wardbell** [comment]
@robwormald  Oh ... now I see what you mean. Yes `XHRBackend` is a class but it _implements_ rather than _extends_ `ConnectionBackend` and the `XHRBackend` has **private** members like `_browserXHR`.

I still don't get how that makes a difference. The `XHRBackend` is still a class that transpiles to a function and that function is clearly in the generated JavaScript which is why it works as a lookup symbol for DI purposes.

Anyway, does this mean that we can't use `InMemoryWebApi` in an AoT solution until that fix?

How do we get that fix queued up?  The `InMemoryWebApi` is an Angular project, FWIW.

**Comment by robwormald** [comment]
its not a runtime problem, its a typechecking problem. The basic issue is that Typescript is comparing `InMemoryWebApi` to  `XHRBackend` - which is wrong (and fails because XHRBackend has an un-overridable implementation (again, per the TS issue)

 It *should* be comparing `InMemoryWebApi` against `ConnectionBackend`, which has zero implementation detail and thus can be implemented properly.

**Comment by wardbell** [comment]
I'm struggling with your suggested workaround because, AFAIK, it requires me to provide a replacement for `Http`. 
1. I don't have such a replacement
2. Wouldn't want to use it if I did; that would mean I prevented the application developer from  substituting his/her own implementation.

I'm missing something crucial about DI. In my experience, it has never mattered that the provider lookup symbol have any relation whatsoever to the provided thing. There is no obligation that `InMemoryWebApi` be compared to `XHRBackend` or anything else. All that matters is that, if something asks for `XHRBackend` , DI should give it an instance of `InMemoryWebApi` .

Apparently AoT changes the rules. I don't understand how but it does.

Seems I am hosed for AoT. And I foresee a whole bunch of other failures of this kind simply because someone told DI "_When something asks for 'A', give them this 'B'_"

I don't see how we can lay this at the doorstep of TypeScript.

**Comment by thelgevold** [comment]
I tried the workaround, but didn't quite get it to work

```
//my custom factory
export function myCustomHttpFactory(options) {
  return (fakeBackend: InMemoryBackendService): Http => {
    return new Http(fakeBackend, options);
  };
}

@NgModule({})
export class InMemoryWebApiModule {
  static forRoot(dbCreator: Type, options?: any) : ModuleWithProviders  {
    return {
      ngModule: InMemoryWebApiModule,
      providers: [
        { provide: Http, useFactory: myCustomHttpFactory(options), deps: [ InMemoryBackendService] },
        { provide: SEED_DATA,  useClass: dbCreator },
        { provide: InMemoryBackendConfig, useValue: options || {} }
      ]
    };
  }
}
```

I think the issue is that factories are not allowed by the AOT compiler. At least it fails with the error:

```
 Error encountered resolving symbol values statically. Function calls are not supported. Consider replacing the function or lambda with a reference to an exported function (position 17:9 in the original
```
This is similar to errors I've seen before when I try to be clever and add logic to the building of the object

**Comment by thelgevold** [comment]
Spoke with @wardbell and there is an alternative factory syntax that I tried instead:
```
export function myCustomHttpFactory(inMem: InMemoryBackendService) {
  return (): Http => {
    return new Http(inMem, null);
  };
}

@NgModule({})
export class InMemoryWebApiModule {
  static forRoot(dbCreator: Type, options?: any) : ModuleWithProviders  {
    return {
      ngModule: InMemoryWebApiModule,
      providers: [
        { provide: Http, useFactory: myCustomHttpFactory, deps: [ InMemoryBackendService] },
        { provide: SEED_DATA,  useClass: dbCreator },
        { provide: InMemoryBackendConfig, useValue: options || {} }
      ]
    };
  }
}
```

This will compile, so I guess we can try this as a workaround for now.

**Comment by robwormald** [comment]
Simple workaround:

```ts
export const tempHttpFactory(backend, requestOptions){
  return new Http(backend, requestOptions)
}

{ provide: Http, useFactory: tempHttpFactory, deps: [ InMemoryWebApi, RequestOptions ] }
```

That's all the current implementation does, except that the signature in the real httpFactory is looking for an XHRBackend - https://github.com/angular/angular/blob/master/modules/%40angular/http/src/http_module.ts#L31

So when the code is generated, it looks something like
```ts
this.http = httpFactory(new InMemoryWebApi(), new RequestOptions())
```
which fails to typecheck, because (as described in the TS issue) InMemoryWebAPI incorrectly implements XHRBackend. The TS team has stated that's a wontfix (and their reasoning is sound, afaict)

Thus, the correct solution is internally, Angular should:
```ts
{ provide: ConnectionBackend, useClass: XHRBackend }
```
and people who want to override should
```ts
{ provide: ConnectionBackend, useClass: MyCustomBackend }
```

The way everybody does it today (overriding XHRBackend) is incorrect, but is the only way to actually do it. The fix is easy but I need to figure out a way to do it without breaking the world.

**Comment by wardbell** [comment]
Thanks, @robwormald , for helping to explain and for at least contemplating a fix before final (hint hint). 

We'll squeak by with a `tempHttpFactory` and hope no one really wants to combine the in-mem backend with their own `Http` provider.

**Comment by alxhub** [comment]
Unfortunately I don't plan on fixing this - people override `Http` in too many different ways to safely change the way it's constructed.

The good news is this won't be an issue in the new HTTP APIs.
